### PR TITLE
Reschedule resident task during a deployment

### PR DESCRIPTION
**Pull request by alenkacz**
Summary:
Current buggy flow:
- start deployment
- launch new resident task
- resident task have failing health check
- resident task is killed because of failing health check
- scale check is triggered but is ignored because we are in the middle of deployment
- `TaskStartActor` handles only `InstanceChanged` of terminal instances so not the reserved-terminal (note: right now we use Reserved for both Terminal and "Scheduled" cases)

Current change means:
- `TaskStartActor` will get `InstanceChanged` event for resident terminal tasks
- if `Sync` kicks in, it will also not consider terminal Reserved instances as active anymore

JIRA issues: DCOS-38317

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/core/deployment/impl/TaskStartActorTest.scala [hunk 1]
```diff
@@ -212,6 +216,31 @@ class TaskStartActorTest extends AkkaUnitTest with Eventually {
       expectTerminated(ref)
     }
   }
+
+  "relaunch when resident task gets terminal" in {
```
Comment (alenkacz) [inline src/test/scala/mesosphere/marathon/core/deployment/impl/TaskStartActorTest.scala]: this test was failing prior to the fix...

Comment (zen-dog) [reply src/test/scala/mesosphere/marathon/core/deployment/impl/TaskStartActorTest.scala]: When did this test fail last time?

Comment (ichernetsky) [reply src/test/scala/mesosphere/marathon/core/deployment/impl/TaskStartActorTest.scala]: I think what Alena means is that this new test fails if the fix in this PR is not applied to the code.

**Comment by zen-dog** [comment]
> We need to follow up with a SI test as well.

I disagree. SI tests are expensive and ideally, represent user feature level. The fix is relatively simple and a unit test to prevent regression seems adequate.

**Comment by timcharper** [comment]
How about an integration test?

**Comment by ichernetsky** [comment]
@zen-dog, it sound like Einstein's quote :)
It needs to be tested more thoroughly, not just by unit-tests, because PVs are one of the most hairy parts of Marathon, and one of the least used parts of Marathon too. On top of that, the cost of error is too high — potentially loosing data. If an integration test will do the trick, I am fine with it.

File: src/test/scala/mesosphere/marathon/core/deployment/impl/TaskStartActorTest.scala (hunk 1 above)
Comment (alenkacz) [reply src/test/scala/mesosphere/marathon/core/deployment/impl/TaskStartActorTest.scala]: Exactly :)

**Comment by alenkacz** [comment]
Yeah, I actually already started working on integration test... These unit tests are so heavily mocked that I personally don't believe it fixes what I think it fixes.

**Pull request merged**